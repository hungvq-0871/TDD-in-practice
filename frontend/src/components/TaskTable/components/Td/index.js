import React from 'react';
import PropTypes from 'prop-types'

import Checkbox from 'components/Checkbox';
import Badge from 'components/Badge';
import TdStyle from './index.style';
import StatusTick from '../StatusTick';

function Td({
  checkboxId,
  title,
  userName,
  time,
  onChange,
}) {
  return (
    <TdStyle>
      <StatusTick />
      <Checkbox id={checkboxId} onChange={onChange} />
      <div className="detail-info">
        <h3 className="title">{title}</h3>
        <div className="created-info">
          <div className="left">
            <span className="created-by">{userName}</span>
            <StatusTick className="tick" />
            <time className="created-at">{time}</time>
          </div>
          <div className="right">
            <Badge color="green">Demo</Badge>
            <Badge color="blue">Meeting</Badge>
            <Badge color="purple">Chil</Badge>
            <Badge color="red">Abcd</Badge>
          </div>
        </div>
      </div>
    </TdStyle>
  )
}

Td.propTypes = {
  checkboxId: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  title: PropTypes.string.isRequired,
  userName: PropTypes.string.isRequired,
  time: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired
}

export default Td;
