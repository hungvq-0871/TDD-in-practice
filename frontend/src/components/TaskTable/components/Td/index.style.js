import styled from 'styled-components';

const TdStyle = styled.td`
  ${({ theme }) => (`
    display: flex;

    .custom-checkbox {
      margin: -3px 10px 0 10px;

      .custom-control-label::before, 
      .custom-control-label::after {
        width: 20px;
        height: 20px;
      }
    }

    .detail-info {
      flex: 1;

      h3 {
        font-size: 20px;
      }

      .created-info {
        display: flex;
        justify-content: space-between;
        color: ${theme.grey1};

        .left,
        .right {
          display: flex;
          align-items: center;
        }

        .tick {
          margin: 0 10px;
          background-color: ${theme.grey1};
        }

        .badge {
          padding: 6px 10px;
          margin-left: 10px;
        }
      }
    }
  `)}
`;

export default TdStyle;