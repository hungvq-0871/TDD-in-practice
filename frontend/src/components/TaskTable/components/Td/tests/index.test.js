import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import Td from '../index';

const renderTdWithProps = (checkboxTd, title, username, time, onChange) => {
  const trEle = document.createElement('tr');
  return render(<Td checkboxId={checkboxTd} title={title} userName={username} time={time} onChange={onChange} />, {
    /*
      *** a "div" ele is the default "container" of all ele rendered by this "render" function. However, since a "td" ele can not be  the child
          of a "div" ele, we have to switch the "container" ele from a "div" ele to a "tr" ele.
    */
    container: document.body.appendChild(trEle),
  });
}

describe('test how "Td" is rendered', () => {
  test(`it's title must match to the value of it's "title" props`, () => {
    const expected = "title";
    const { container } = renderTdWithProps(1, expected, "username", "10 PM", () => { });
    const titleEle = container.querySelector('.title');

    expect(titleEle.innerHTML).toBe(expected);
  });

  test(`it's created by indicator must match to the value of it's "username" props`, () => {
    const expected = "username";
    const { container } = renderTdWithProps(1, "title", expected, "10 PM", () => { });
    const createdByEle = container.querySelector('.created-by');

    expect(createdByEle.innerHTML).toBe(expected);
  });

  test(`it's created at (time) indicator must match to the value of it's "username" props`, () => {
    const expected = "10 PM";
    const { container } = renderTdWithProps(1, "title", "name", expected, () => { });
    const createdAtEle = container.querySelector('.created-at');

    expect(createdAtEle.innerHTML).toBe(expected);
  });

  test(`it should match it's snapshot`, () => {
    const { container } = renderTdWithProps(1, "title", "name", "10 PM", () => { });
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('test functions around "Td"', () => {
  test(`it should match it's snapshot`, () => {
    const handleOnChange = jest.fn(() => { });
    const { container } = renderTdWithProps(1, "title", "name", "10 PM", handleOnChange);
    const checkboxEle = container.querySelector(".custom-control-label");

    fireEvent.click(checkboxEle);

    expect(handleOnChange.mock.calls.length).toBe(1);
  });
});
