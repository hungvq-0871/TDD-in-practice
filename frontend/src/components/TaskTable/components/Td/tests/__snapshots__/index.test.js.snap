// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test how "Td" is rendered it should match it's snapshot 1`] = `
<td
  class="sc-bwzfXH XJyym"
>
  <span
    class="sc-htpNat lgZCgg"
  />
  <div
    class="custom-control custom-checkbox"
  >
    <input
      class="custom-control-input"
      id="1"
      type="checkbox"
      value=""
    />
    <label
      class="custom-control-label"
      for="1"
    />
  </div>
  <div
    class="detail-info"
  >
    <h3
      class="title"
    >
      title
    </h3>
    <div
      class="created-info"
    >
      <div
        class="left"
      >
        <span
          class="created-by"
        >
          name
        </span>
        <span
          class="sc-htpNat lgZCgg tick"
        />
        <time
          class="created-at"
        >
          10 PM
        </time>
      </div>
      <div
        class="right"
      >
        <span
          class="sc-bdVaJa hntilx badge badge-green badge-pill"
        >
          Demo
        </span>
        <span
          class="sc-bdVaJa hntilx badge badge-blue badge-pill"
        >
          Meeting
        </span>
        <span
          class="sc-bdVaJa hntilx badge badge-purple badge-pill"
        >
          Chil
        </span>
        <span
          class="sc-bdVaJa hntilx badge badge-red badge-pill"
        >
          Abcd
        </span>
      </div>
    </div>
  </div>
</td>
`;
